src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -0,0 +1,157 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Globalization;
+
+namespace System.Text.Json
+{
+    internal abstract class JsonSeparatorNamingPolicy : JsonNamingPolicy
+    {
+        private readonly bool _lowercase;
+        private readonly char _separator;
+
+        internal JsonSeparatorNamingPolicy(bool lowercase, char separator) =>
+            (_lowercase, _separator) = (lowercase, separator);
+
+        public sealed override string ConvertName(string name)
+        {
+            // Rented buffer 20% longer that the input.
+            int bufferLength = (12 * name.Length) / 10;
+            char[]? buffer = bufferLength > JsonConstants.StackallocCharThreshold
+                ? ArrayPool<char>.Shared.Rent(bufferLength)
+                : null;
+
+            int resultLength = 0;
+            Span<char> result = buffer is null
+                ? stackalloc char[JsonConstants.StackallocCharThreshold]
+                : buffer;
+
+            void ExpandBuffer(ref Span<char> result)
+            {
+                var bufferNew = ArrayPool<char>.Shared.Rent(bufferLength *= 2);
+
+                result.CopyTo(bufferNew);
+
+                if (buffer is not null)
+                {
+                    ArrayPool<char>.Shared.Return(buffer, clearArray: true);
+                }
+
+                buffer = bufferNew;
+                result = buffer;
+            }
+
+            void WriteWord(ReadOnlySpan<char> word, ref Span<char> result)
+            {
+                if (word.IsEmpty)
+                {
+                    return;
+                }
+
+                Span<char> destination = result.Slice(resultLength != 0
+                    ? resultLength + 1
+                    : resultLength);
+
+                int written;
+                while (true)
+                {
+                    written = _lowercase
+                        ? word.ToLowerInvariant(destination)
+                        : word.ToUpperInvariant(destination);
+
+                    if (written > 0)
```

Comment:
Do we need to call `ToLowerInvariant` to determine that `destination` has enough space?  Isn't it the case that we could simply check if `destination.Length >= word.Length` and grow the buffer _before_ attempting to make a conversion?

Reply:
Let's no rely on internals of text conversion. While you may be correct, and that was in the original code, I would not use that approach because of potential bugs.

Reply:
See discussion here https://github.com/dotnet/runtime/pull/69613#discussion_r940307432

Reply:
https://github.com/dotnet/runtime/pull/69613#discussion_r993553302